BUG: Common NumericIndex.__new__, with fixed name handling

toobaz opened the pull request:
closes #12309

toobaz committed:
BUG: Common NumericIndex.__new__, fixed name handling in indices

closes #12309

jreback commented on pandas/indexes/numeric.py:
```diff
@@ -19,6 +19,32 @@ class NumericIndex(Index):
+
+        if copy or data.dtype != cls._default_dtype:
+            try:
+                subarr = np.array(data, cls._default_dtype, copy=copy)
```
pass ``dtype`` by kw

jreback commented on pandas/indexes/numeric.py:
```diff
@@ -19,6 +19,32 @@ class NumericIndex(Index):
+        if copy or data.dtype != cls._default_dtype:
+            try:
+                subarr = np.array(data, cls._default_dtype, copy=copy)
+                assert((subarr == data) | np.isnan(subarr)).all()
```
what is the purpose of this check?

toobaz replied:
I mostly copied from ``Int64Index.__new__``, where it guarantees that

        Int64Index([1.2, 3])

fails, and adapted to work also for ``Float64Index``, where... it is probably useless.

jreback replied:
ok, will need several tests w.r.t. to whether a passed underlying array is actually copied or not (there are some already in ``indexes/tests/test_numeric.py/Numeric`` but have to make sure that these are correctly tested

jreback commented:
pls add tests for the name propogation

toobaz committed:
TST: Added test for copying of data

toobaz commented:
Ping?
(in general, are you postponing non-0.18.0-tagged PRs to after the 0.18.0 release?)

jreback commented:
@toobaz yes

jreback commented:
though non-api changes or straightforward stuff I will merge (so pinging is good)

jreback commented on pandas/indexes/multi.py:
```diff
@@ -66,8 +66,10 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
-        if name is not None:
-            names = name
+        if name is None and hasattr(levels, 'name'):
+            name = levels.name
```
how does this check help? ``levels`` is a list-like

jreback commented on pandas/indexes/multi.py:
```diff
@@ -66,8 +66,10 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
+        if name is None and hasattr(levels, 'name'):
+            name = levels.name
+        if isinstance(levels, MultiIndex):
+            return levels.copy(name=name, names=names, deep=copy)
```
*this* could also be true I agree, but then it should be *before* the prior check, yes?
I also really don't like the ``name``, ``names`` biz. I believe ``Index`` has the right way to do this (IOW check for both ``name`` and ``names`` if both are specified its an error, otherwise use the non ``None`` one)

toobaz replied:
I'm not really following you here... where does ``Index`` do this?

jreback commented on pandas/indexes/multi.py:
```diff
@@ -344,6 +344,7 @@ def copy(self, names=None, dtype=None, levels=None, labels=None,
         dtype : numpy dtype or pandas type, optional
         levels : sequence, optional
         labels : sequence, optional
+        name : object, optional
```
this shoul dlook much like the ``Index`` one (as far as the name checking goes)

jreback commented on pandas/indexes/numeric.py:
```diff
@@ -19,6 +19,32 @@ class NumericIndex(Index):
     """
     _is_numeric_dtype = True
 
+    def __new__(cls, data=None, dtype=None, copy=False, name=None,
```
I dont' think we need to accept ``kwargs`` here (or if we do, then use use something like ``core.generic._validate_kwargs`` to allow certain compat keywords (but then we don't advertise them)

(note the location/name of this is changing, but ok for now)

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -654,3 +654,19 @@ def test_fillna(self):
+            first = index.__class__(index, copy=True, name='mario')
+            second = first.__class__(first, copy=False)
+            self.assertTrue(index.equals(first))
+            # Even though "copy=False", we want a new object:
```
blank line

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -654,3 +654,19 @@ def test_fillna(self):
+            second = first.__class__(first, copy=False)
+            self.assertTrue(index.equals(first))
+            # Even though "copy=False", we want a new object:
+            self.assertTrue(id(first) != id(second))
```
use ``.identical(...)``

toobaz replied:
In place of ``.equals()``?! But the name is different...

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -654,3 +654,19 @@ def test_fillna(self):
+            self.assertTrue(id(first) != id(second))
+
+            if isinstance(index, MultiIndex) and len(index.levels) > 1:
+                # No unique "name" attribute (each level has its own)
```
override this for the ``MultiIndex`` tests

jreback replied:
and just skip the ``MultIndex`` in the iteration (as you need a more comprehensive MI test for copy)

toobaz replied:
What do you mean by "override"/"skip"?

jreback commented on pandas/tests/indexes/test_numeric.py:
```diff
@@ -831,3 +831,11 @@ def test_ufunc_coercions(self):
         tm.assert_index_equal(result, exp)
+
+    def test_ensure_copied_data(self):
+        data = np.array([1, 2, 3], dtype='int64')
```
add the issue number

jreback commented on pandas/tests/indexes/test_numeric.py:
```diff
@@ -831,3 +831,11 @@ def test_ufunc_coercions(self):
+        data = np.array([1, 2, 3], dtype='int64')
+        idx = Int64Index(data, copy=True)
+        self.assert_numpy_array_equal(data, idx._data)
+        tm.assertIsNot(data, idx._data)
```
this is a VERY specific test, can you make this much more general (e.g. use ``.create_data()``) to ensure that this is correct for numeric Index types. Furthermore, would be ideal to make this test for ALL Indexes (IOW put this in TestBase)

jreback commented on pandas/tests/test_common.py:
```diff
@@ -725,9 +725,9 @@ def test_ensure_platform_int():
+    # int32 - "dtype" argument is irrelevant
     x = Int64Index([1, 2, 3], dtype='int32')
-    assert (x.dtype == np.int32)
+    assert (x.dtype == np.int64)
```
I think let's raise on a passed ``dtype``, see what breaks (raise a ``TypeError``)

jreback commented:
@toobaz if you'd like to rebase / update, pls reopen.

toobaz commented:
Rebased, updated, addressed your comments, except a couple of them which were unclear to me. I don't think I am allowed to reopen.

While extending ``est_ensure_copied_data`` as you suggested, a bug came out in ``DateTimeIndex``, now fixed.

toobaz commented:
(not sure why the code isn't being updated... shall I open a new PR?)

jorisvandenbossche commented:
@toobaz yes, you will have to open a new PR. That's something with github that you cannot reopen a PR once the branch has been updated while the PR is closed (so we should have reopened it before you pushed updated to the branch ..). Anyway, no problem to open a new one! :-)
